In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

df_original = pd.read_csv('BankChurners.csv')
dropColumns = ['Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1', 
               'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2']
df = df_original.drop(columns=dropColumns)
df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Months_Inactive_12_mon  Contacts_Count_12_mon  \
0              39  ...                       1                      3   
1              44  ...                       1                      2   
2              36  ...                       1                      0   
3              34  ...                       4                      1   
4              21  ...                       1                      0   

   Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  \
0       12691.0                  777          11914.0                 1.335   
1        8256.0                  864           7392.0                 1.541   
2        3418.0                    0           3418.0                 2.594   
3        3313.0                 2517            796.0                 1.405   
4        4716.0                    0           4716.0                 2.175   

   Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  
0             1144              42                1.625                  0.061  
1             1291              33                3.714                  0.105  
2             1887              20                2.333                  0.000  
3             1171              20                2.333                  0.760  
4              816              28                2.500                  0.000  

[5 rows x 21 columns]

In [10]:
df['Dependent_count'] = df['Dependent_count'].astype(str)
df_dummies = pd.get_dummies(df, drop_first=True)
df_dummies.head()

CLIENTNUM  Customer_Age  Months_on_book  Total_Relationship_Count  \
0  768805383            45              39                         5   
1  818770008            49              44                         6   
2  713982108            51              36                         4   
3  769911858            40              34                         3   
4  709106358            40              21                         5   

   Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
0                       1                      3       12691.0   
1                       1                      2        8256.0   
2                       1                      0        3418.0   
3                       4                      1        3313.0   
4                       1                      0        4716.0   

   Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  ...  \
0                  777          11914.0                 1.335  ...   
1                  864           7392.0                 1.541  ...   
2                    0           3418.0                 2.594  ...   
3                 2517            796.0                 1.405  ...   
4                    0           4716.0                 2.175  ...   

   Marital_Status_Single  Marital_Status_Unknown  Income_Category_$40K - $60K  \
0                      0                       0                            0   
1                      1                       0                            0   
2                      0                       0                            0   
3                      0                       1                            0   
4                      0                       0                            0   

   Income_Category_$60K - $80K  Income_Category_$80K - $120K  \
0                            1                             0   
1                            0                             0   
2                            0                             1   
3                            0                             0   
4                            1                             0   

   Income_Category_Less than $40K  Income_Category_Unknown  \
0                               0                        0   
1                               1                        0   
2                               0                        0   
3                               1                        0   
4                               0                        0   

   Card_Category_Gold  Card_Category_Platinum  Card_Category_Silver  
0                   0                       0                     0  
1                   0                       0                     0  
2                   0                       0                     0  
3                   0                       0                     0  
4                   0                       0                     0  

[5 rows x 38 columns]

In [11]:
df_dummies = df_dummies.drop(['CLIENTNUM'], axis=1)

In [12]:
columns = df_dummies.columns.values
columns

array(['Customer_Age', 'Months_on_book', 'Total_Relationship_Count',
       'Months_Inactive_12_mon', 'Contacts_Count_12_mon', 'Credit_Limit',
       'Total_Revolving_Bal', 'Avg_Open_To_Buy', 'Total_Amt_Chng_Q4_Q1',
       'Total_Trans_Amt', 'Total_Trans_Ct', 'Total_Ct_Chng_Q4_Q1',
       'Avg_Utilization_Ratio', 'Attrition_Flag_Existing Customer',
       'Gender_M', 'Dependent_count_1', 'Dependent_count_2',
       'Dependent_count_3', 'Dependent_count_4', 'Dependent_count_5',
       'Education_Level_Doctorate', 'Education_Level_Graduate',
       'Education_Level_High School', 'Education_Level_Post-Graduate',
       'Education_Level_Uneducated', 'Education_Level_Unknown',
       'Marital_Status_Married', 'Marital_Status_Single',
       'Marital_Status_Unknown', 'Income_Category_$40K - $60K',
       'Income_Category_$60K - $80K', 'Income_Category_$80K - $120K',
       'Income_Category_Less than $40K', 'Income_Category_Unknown',
       'Card_Category_Gold', 'Card_Category_Platinum',
    

In [13]:
index = np.where(columns == 'Attrition_Flag_Existing Customer')[0]
X_indexes = list(np.delete(columns, index))
y_indexes = ['Attrition_Flag_Existing Customer']

X = df_dummies[X_indexes]
y = df_dummies[y_indexes]

X_standarized = StandardScaler().fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X_standarized, y, test_size = .30, random_state=42)

clf = LogisticRegression(random_state = 0, solver='lbfgs', max_iter=200).fit(X_train, y_train)
y_pred = clf.predict(X_test)
clf.score(X_test, y_test)

C:\Users\cmdra\anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


0.8983218163869694

In [14]:
cf = confusion_matrix(y_test, y_pred)
cf

array([[ 271,  225],
       [  84, 2459]], dtype=int64)

This model has an accuracy of 89.83%. In the confusion matrix we can see the majority of the errors are type 1 and the errors type 2 are quite low. This model is pretty decent, can be improved but this is not the current goal. We need to check the coefficients of the model in order to see which variables are more realated to the churn action.

In [15]:
coef = pd.DataFrame(clf.coef_[0], index = X_indexes, columns = ['Coefficients'])
coef

Coefficients
Customer_Age                       -0.053592
Months_on_book                      0.101826
Total_Relationship_Count            0.682966
Months_Inactive_12_mon             -0.524236
Contacts_Count_12_mon              -0.586109
Credit_Limit                        0.109766
Total_Revolving_Bal                 0.716785
Avg_Open_To_Buy                     0.045483
Total_Amt_Chng_Q4_Q1                0.104940
Total_Trans_Amt                    -1.559461
Total_Trans_Ct                      2.714303
Total_Ct_Chng_Q4_Q1                 0.624902
Avg_Utilization_Ratio               0.109695
Gender_M                            0.484580
Dependent_count_1                   0.015353
Dependent_count_2                  -0.122074
Dependent_count_3                  -0.142013
Dependent_count_4                  -0.091061
Dependent_count_5                  -0.132177
Education_Level_Doctorate          -0.118593
Education_Level_Graduate           -0.082850
Education_Level_High School        -0.048673
Education_Level_Post-Graduate      -0.082702
Education_Level_Uneducated         -0.083152
Education_Level_Unknown            -0.125458
Marital_Status_Married              0.275911
Marital_Status_Single              -0.005297
Marital_Status_Unknown              0.048482
Income_Category_$40K - $60K         0.323156
Income_Category_$60K - $80K         0.131314
Income_Category_$80K - $120K        0.072517
Income_Category_Less than $40K      0.328998
Income_Category_Unknown             0.217252
Card_Category_Gold                 -0.119149
Card_Category_Platinum             -0.027208
Card_Category_Silver               -0.055438

In [16]:
no_dummie_coef = coef.iloc[0:13, 0]
dummie_coef = coef.iloc[13:len(coef), 0]

In [17]:
dummie_coef.sort_values()

Dependent_count_3                -0.142013
Dependent_count_5                -0.132177
Education_Level_Unknown          -0.125458
Dependent_count_2                -0.122074
Card_Category_Gold               -0.119149
Education_Level_Doctorate        -0.118593
Dependent_count_4                -0.091061
Education_Level_Uneducated       -0.083152
Education_Level_Graduate         -0.082850
Education_Level_Post-Graduate    -0.082702
Card_Category_Silver             -0.055438
Education_Level_High School      -0.048673
Card_Category_Platinum           -0.027208
Marital_Status_Single            -0.005297
Dependent_count_1                 0.015353
Marital_Status_Unknown            0.048482
Income_Category_$80K - $120K      0.072517
Income_Category_$60K - $80K       0.131314
Income_Category_Unknown           0.217252
Marital_Status_Married            0.275911
Income_Category_$40K - $60K       0.323156
Income_Category_Less than $40K    0.328998
Gender_M                          0.484580
Name: Coeff

It looks like the most important feature from the dummies variables is the Gender. We can see that being male impacts positively on Existing client which means that the females are more prone to churn the bank, this is something that we already measured. We also can see that the people with lower income are more prone to stay in the bankin the same way people with status Married, and people who don't declare the income.

In [18]:
no_dummie_coef.sort_values()

Total_Trans_Amt            -1.559461
Contacts_Count_12_mon      -0.586109
Months_Inactive_12_mon     -0.524236
Customer_Age               -0.053592
Avg_Open_To_Buy             0.045483
Months_on_book              0.101826
Total_Amt_Chng_Q4_Q1        0.104940
Avg_Utilization_Ratio       0.109695
Credit_Limit                0.109766
Total_Ct_Chng_Q4_Q1         0.624902
Total_Relationship_Count    0.682966
Total_Revolving_Bal         0.716785
Total_Trans_Ct              2.714303
Name: Coefficients, dtype: float64

This variables are a bit more complicate to interpret due to the difference between measures units. 
 - We can see that the 'Total_Trans_Ct' has a positive inpact on the client to stay and this impact seems to be quite significant. This feature measure the total number of transactions made during the last 12 moths. So this means that clients which make more transactions during the last 12 months are more prone to stay which make sense.
 - We can also see that the 'Total_Trans_Amt' has a significant negative impact on the decision to stay of the client. This feature is the total transactio amount during the last 12 months. In certain way this also make sense because when a client churn the bank will take out all theis saves.
 - We also can see on 'Total_Revolving_Bal' that a client with higher depts is more prone to stay.
 - With the feature 'Total_Relationship_Count' we can see that the more products a client hire the more prone is to stay.
 - With the feature 'Total_Ct_Chng_Q4_Q1' we can see that the more they spent in Q4 regarding to Q1 the more prone are to stay.
 - With the feature 'Contacts_Count_12_mon' we can see that the more the client contact the bank the more prone are to curn it.
 - With the feature 'Months_Inactive_12_mon' we can see that the more inactive is the client the more prone is to churn the company.